### DISCUSSION TITLE
client websocket

### DISCUSSION THREAD

**Discussion by KH93b** [nanoexpress/nanoexpress#455] [General]
which client is better to use for websocket server?

**Comment by dalisoft** [comment]
Hi @KH93b 
I always use native browser websocket client with manual reconnect logic

**Reply by KH93b** [reply]
Thanks for the answer. I can't figure out how to get data from the session in the socket to identify the user. Do you have an example of this.
I don't want the connection to be established if the user is not logged in.
I would be very grateful for a small example.

**Reply by dalisoft** [reply]
https://github.com/nanoexpress/nanoexpress/blob/master/examples/websocket.js#L8

user session should be within `req` and then you can use after `connection` event

**Reply by KH93b** [reply]
I get the session via http, and the websocket session is undefined.

**Reply by dalisoft** [reply]
@KH93b Are you tried PRO Slim or Ultimate version? These versions also exist this bug?

**Reply by KH93b** [reply]
I did not know about these versions. Which one is better for production use?

**Reply by KH93b** [reply]
I looked at ultimate, compared... Yeah, that's cool. I'll try to use it in my free project.

**Reply by dalisoft** [reply]
@KH93b the best one is default nanoexpress because of free license. Not locked or not strictly locks you, you can use it on all your projects. 

Ultimate version is feature packed and a little faster but you cannot use it on your private projects due of GPL-3.0 license. Only on open source projects. 
if you want use it on your private project, you should be sponsor of project

**Reply by KH93b** [reply]
I'm doing a school project, just a project for me. Can I learn the Ultimate version on it, or will some features be blocked?

**Reply by dalisoft** [reply]
You can search about GPL-3.0 license. 
tldr: if your project is not open source, you should not use ultimate version unless you are project sponsor

**Reply by KH93b** [reply]
Yes, this is an open source educational project.

**Reply by dalisoft** [reply]
Then sure you can use if project open source
